### PR TITLE
mgr/dashboard: add ssl verify option for prometheus and alert manager

### PR DESCRIPTION
**Pull request by henyxia**
Fixes: https://tracker.ceph.com/issues/47863

Signed-off-by: Jean "henyxia" Wasilewski <henyxia@revs0.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/prometheus.py [hunk 1]
```diff
@@ -30,20 +30,22 @@ class PrometheusRESTController(RESTController):
     def prometheus_proxy(self, method, path, params=None, payload=None):
         # type (str, str, dict, dict)
         return self._proxy(self._get_api_url(Settings.PROMETHEUS_API_HOST),
-                           method, path, 'Prometheus', params, payload)
+                           method, path, 'Prometheus', params, payload,
+                           Settings.PROMETHEUS_API_SSL_VERIFY)
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/prometheus.py:34]: Minor thing: just for the sake of clarity (a method with 7 positional arguments starts being hard to follow), I'd add this a as keyword argument:
```suggestion
                           verify=Settings.PROMETHEUS_API_SSL_VERIFY)
```

**Review by epuertat — COMMENTED** [review]
Thanks for your contribution, @henyxia !

It seems this is [not passing unit tests](https://jenkins.ceph.com/job/ceph-pull-requests/61697/):
```py
TypeError: the JSON object must be str, bytes or bytearray, not 'MagicMock'
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/services/exception.py", line 46, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/.tox/py3/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 60, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/__init__.py", line 677, in inner
    ret = func(*args, **kwargs)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/__init__.py", line 872, in wrapper
    return func(*vpath, **params)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/prometheus.py", line 78, in create_silence
    return self.alert_proxy('POST', '/silences', payload=params)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/prometheus.py", line 40, in alert_proxy
    Settings.ALERTMANAGER_API_SSL_VERIFY)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/prometheus.py", line 54, in _proxy
    content = json.loads(response.content)
  File "/usr/lib/python3.6/json/__init__.py", line 348, in loads
    'not {!r}'.format(s.__class__.__name__))
TypeError: the JSON object must be str, bytes or bytearray, not 'MagicMock'

```

**Comment by henyxia** [comment]
> Thanks for your contribution, @henyxia !
> 
> It seems this is [not passing unit tests](https://jenkins.ceph.com/job/ceph-pull-requests/61697/):
> 
> ```python
> TypeError: the JSON object must be str, bytes or bytearray, not 'MagicMock'
> Traceback (most recent call last):
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/services/exception.py", line 46, in dashboard_exception_handler
>     return handler(*args, **kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/.tox/py3/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 60, in __call__
>     return self.callable(*self.args, **self.kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/__init__.py", line 677, in inner
>     ret = func(*args, **kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/__init__.py", line 872, in wrapper
>     return func(*vpath, **params)
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/prometheus.py", line 78, in create_silence
>     return self.alert_proxy('POST', '/silences', payload=params)
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/prometheus.py", line 40, in alert_proxy
>     Settings.ALERTMANAGER_API_SSL_VERIFY)
>   File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/prometheus.py", line 54, in _proxy
>     content = json.loads(response.content)
>   File "/usr/lib/python3.6/json/__init__.py", line 348, in loads
>     'not {!r}'.format(s.__class__.__name__))
> TypeError: the JSON object must be str, bytes or bytearray, not 'MagicMock'
> ```

Yes, I will try to fix this tonight or tomorrow and submit it again!
Thank you :)

**Comment by henyxia** [comment]
Well, tests are working properly now
But I did not manage to add test cases where verify options are set to false
Do you think it should be sufficient or do you want me to add more test? Or add them later in another MR?

**Comment by henyxia** [comment]
@epuertat I created an issue on the tracker and edited the commit message to reflect this
https://tracker.ceph.com/issues/47863

**Comment by epuertat** [comment]
> @epuertat I created an issue on the tracker and edited the commit message to reflect this
> https://tracker.ceph.com/issues/47863

@henyxia thanks a lot! One question: if this is for using with self-signed certificates, there's a [better approach](https://requests.readthedocs.io/en/master/user/advanced/#ssl-cert-verification):
```py
        PROMETHEUS_API_SSL_VERIFY = '/path/to/self/signed/certificate.pem'
```
This will actually ensure reliable server-side authentication.

**Comment by henyxia** [comment]
> > @epuertat I created an issue on the tracker and edited the commit message to reflect this
> > https://tracker.ceph.com/issues/47863
> 
> @henyxia thanks a lot! One suggestion: if this is for using with self-signed certificates, there's a [better approach](https://requests.readthedocs.io/en/master/user/advanced/#ssl-cert-verification):
> 
> ```python
>         PROMETHEUS_API_SSL_VERIFY = '/path/to/self/signed/certificate.pem'
> ```
> 
> This will actually ensure reliable server-side authentication.

Yes this will indeed provide a better security approach.
However, would you agree to add it in another MR?
So user will have the choice to either disable SSL verification or verify it with a custom certificate.

**Review by epuertat — APPROVED** [review]
LGTM! Thanks a lot, @henyxia !

**Comment by LenzGr** [comment]
Would it make sense to include an update to the [Dashboard documentation](https://docs.ceph.com/en/latest/mgr/dashboard/#enabling-prometheus-alerting) along with this PR? New config options should be documented. Thanks!

**Review by votdev — APPROVED** [review]

**Comment by henyxia** [comment]
@LenzGr documentation added :)

**Comment by bk201** [comment]
QA results:
- [CI Branch](https://github.com/ceph/ceph-ci/tree/wip-kiefer-testing-2020-10-27-1440)
- [Package Build](https://github.com/ceph/ceph-ci/tree/wip-kiefer-testing-2020-10-27-1440)
- [Jobs (18 passed / 18 total)](https://pulpito.ceph.com/kchang-2020-10-27_08:21:10-rados:dashboard-wip-kiefer-testing-2020-10-27-1440-distro-basic-smithi/)

**Review by p-se — APPROVED** [review]
lgtm

**Comment by LenzGr** [comment]
Thanks a lot for your contribution, @henyxia !
